assets/dev/js/frontend/utils/anchor-scroll-margin.js

```diff
@@ -0,0 +1,153 @@
+module.exports = elementorModules.ViewModule.extend( {
+	getDefaultSettings() {
+		return {
+			selectors: {
+				links: '.elementor-element a[href*="#"]',
+				stickyElements: '.elementor-element.elementor-sticky',
+			},
+		};
+	},
+
+	onInit() {
+		const that = this;
+
+		this.observeStickyElements( () => {
+			that.initializeStickyAndAnchorTracking();
+		} );
+	},
+
+	observeStickyElements( callback ) {
+		const observer = new MutationObserver( ( mutationsList ) => {
+			for ( const mutation of mutationsList ) {
+				if ( 'childList' === mutation.type || ( 'attributes' === mutation.type && mutation.target.classList.contains( 'elementor-sticky' ) ) ) {
+					callback();
+				}
+			}
+		} );
+
+		observer.observe( document.body, {
+			childList: true,
+			subtree: true,
+			attributes: true,
+			attributeFilter: [ 'class', 'style' ],
+		} );
+	},
+
+	initializeStickyAndAnchorTracking() {
+		const anchorLinks = this.getAllAnchorLinks();
+		const stickyElements = this.getAllStickyElements();
+		const trackedElements = [];
+
+		if ( stickyElements.length > 0 ) {
+			this.trackStickyElements( stickyElements, trackedElements );
+		}
+
+		if ( anchorLinks.length > 0 ) {
+			this.trackAnchorLinks( anchorLinks, trackedElements );
+		}
+
+		this.organizeStickyAndAnchors( trackedElements );
+	},
+
+	trackAnchorLinks( anchorLinks, trackedElements ) {
+		anchorLinks.forEach( ( element ) => {
+			const target = this.getAnchorTarget( element );
+			const scrollPosition = this.getScrollPosition( target );
+			trackedElements.push( {
+				element: target,
+				type: 'anchor',
+				scrollPosition,
+			} );
+		} );
+	},
+
+	trackStickyElements( stickyElements, trackedElements ) {
+		stickyElements.forEach( ( element ) => {
+			const { sticky_anchor_link_offset: scrollMarginTop } = this.getElementSettings( element );
+			if ( 0 === scrollMarginTop ) {
+				return;
+			}
+			const scrollPosition = this.getScrollPosition( element );
+			trackedElements.push( {
+				scrollMarginTop,
+				type: 'sticky',
+				scrollPosition,
+			} );
+		} );
+	},
+
+	organizeStickyAndAnchors( elements ) {
+		const stickyList = this.filterAndSortElementsByType( elements, 'sticky' );
+		const anchorList = this.filterAndSortElementsByType( elements, 'anchor' );
+
+		stickyList.forEach( ( sticky, index ) => {
+			this.defineCurrentStickyRange( sticky, index, stickyList, anchorList );
+		} );
+	},
+
+	defineCurrentStickyRange( sticky, index, stickyList, anchorList ) {
+		const nextStickyScrollPosition = ( index + 1 < stickyList.length )
+			? stickyList[ index + 1 ].scrollPosition
+			: Infinity;
+		sticky.anchor = anchorList.filter( ( anchor ) => {
+			const withinRange = anchor.scrollPosition > sticky.scrollPosition && anchor.scrollPosition < nextStickyScrollPosition;
+			if ( withinRange ) {
+				anchor.element.style.scrollMarginTop = `${ sticky.scrollMarginTop }px`;
```

Comment:
Do we need to handle an offset value for the admin bar well? The admin bar adds something like 32px to the frontend, if it is visible. I assume that we will need to add that value conditionally.

Reply:
@hein-obox do you have a reference for implementing this

Reply:
@Batsirai-muchareva 
https://github.com/elementor/elementor/blob/9eeee9b0695700187cda24b97962a726c80a47cb/assets/dev/js/frontend/utils/anchors.js#L54